### PR TITLE
Added TFeynman class with basic functionality

### PR DESCRIPTION
**Pull request by AdvaitDhingra**
As discussed in the Mattermost groupchat, I have made a class for making Feynman Diagrams. The basic functionality is there, it can be improved in the future. I re-created the feynman.C example in the tutorials/graphics folder using about 1/5th the amount of code. 

Please let me know what you think.

King Regards,

Advait Dhingra

### CODE REVIEW

**Comment by couet** [comment]
Nice work  the macro example is more compact than the original one. But :

- There is no documentation. What is the meaning of the parameter of each method ? You need a header explaining how to use the class.
- You added the new. class in a dedicated folder may be it should go bin graphics.
- How does it compare with exiting tools like the TikZ one mentioned in mattermost ?

**Comment by AdvaitDhingra** [comment]
@couet That you for your feedback. I will add some documentation and move the folder to bin/graphics.

My thought process was that this could potentially be better than TikZ, as this is built completely using ROOT components and it can be changed over time. This class is also quite easy for beginners to contribute to, which would encourage more people to contribute.

**Comment by AdvaitDhingra** [comment]
> * You added the new. class in a dedicated folder may be it should go bin graphics.

I could not find this location. Is it in /bindings?

**Comment by couet** [comment]
Sorry I meant grad2d . You do not need a README.md ... just two file TFeynman.cxx and TFeynman.h .

**Comment by AdvaitDhingra** [comment]
> Sorry I meant grad2d . You do not need a README.md ... just two file TFeynman.cxx and TFeynman.h .

Ah ok, I will move the folder and remove the README and the test.C files. An example is provided at the top of TFeynman.cxx

**Comment by AdvaitDhingra** [comment]
How can I fix the issue that the deepcode-ci-bot is warning me of?

File: graf2d/feynman/inc/TFeynman.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+			TCanvas *c1 = new TCanvas("c1", "c1", 10,10, canvasWidth, canvasHeight);
+   			c1->Range(0, 0, 140, 60);
+			gStyle->SetLineWidth(2);
+		}
+
+		// Fermions:
+		// Quarks:
+        // plots a Quark
+		void Quark(
+            Double_t x1, // x-Coordinate of starting point
```
Comment (couet) [inline graf2d/feynman/inc/TFeynman.h:32]: The data members should be documented the Doxygen way. see https://root.cern/for_developers/doxygen/#data-member-documentation

File: graf2d/feynman/inc/TFeynman.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            Double_t y1, // coordinates of center of the circle
+            Double_t rad, // radius of the circle
+            Double_t labelPositionX, //
+            Double_t labelPositionY, // coordinates of the circle
+            const char * whichLepton, // name of the first lepton
+            const char * whichAntiLepton // name of the antilepton);
+        ); 
+		
+		// Bosons: 
+        // Guage Bosons:
```
Comment (couet) [inline graf2d/feynman/inc/TFeynman.h:73]: Gauge ?

File: graf2d/feynman/src/TFeynman.cxx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  f->CurvedPhoton(30, 30, 12.5*sqrt(2), 135, 225, 7, 30);
+  f->Photon(30, 30, 55, 30, 42.5, 37.7);
+  f->QuarkAntiQuark(70, 30, 15, 55, 45, "q");
+  f->Gluon(70, 45, 70, 15, 77.5, 30);
+  f->WeakBoson(85, 30, 110, 30, 100, 37.5, "Z^{0}");
+  f->Quark(110, 30, 130, 50, 135, 55, "q", true);
+  f->Quark(110, 30, 130, 10, 135, 6, "q", false);
+  f->CurvedGluon(110, 30, 12.5*sqrt(2), 315, 45, 135, 30);
+~~~
+No drawing is needed, as this is all done in the method itself.
```
Comment (couet) [inline graf2d/feynman/src/TFeynman.cxx:29]: "now drawing needed" is against the ROOT philosophy. And the objects created by ROOT needed to be 1st created then 2ndly, Drawn...

File: graf2d/feynman/src/TFeynman.cxx [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        }
+    }
+    TLatex t;
+    t.SetTextSize(0.1);
+    t.DrawLatex(labelPositionX, labelPositionY, usedQuarkName);
+
+    q->Draw();
+}
+
+void TFeynman::QuarkAntiQuark(Double_t x1, Double_t y1, Double_t rad, Double_t labelPositionX, Double_t labelPositionY, const char * quarkName) {
```
Comment (couet) [inline graf2d/feynman/src/TFeynman.cxx:109]: Youmethod are not documented .... see https://root.cern/for_developers/doxygen/#method-documentation

File: graf2d/feynman/inc/TFeynman.h (hunk 2 above)
Comment (AdvaitDhingra) [reply graf2d/feynman/inc/TFeynman.h:73]: In particle physics, a gauge boson is a force carrier, a bosonic particle that carries any of the fundamental interactions of nature, commonly called forces. Elementary particles, whose interactions are described by a gauge theory, interact with each other by the exchange of gauge bosons—usually as virtual particles. I misspelt it here, sorry 😅

File: graf2d/feynman/src/TFeynman.cxx (hunk 4 above)
Comment (AdvaitDhingra) [reply graf2d/feynman/src/TFeynman.cxx:109]: Okay, I will fix this

**Comment by AdvaitDhingra** [comment]
All of the checks seem to have passed

**Comment by couet** [comment]
I see you put the class directly in graf2d. I would not add the feynman folder but rather put the file in graf2d/graf/src and graf2d/graf/inc . You need then to modify graf2d/graf/inc/LinkDef.h

**Comment by AdvaitDhingra** [comment]
> I see you put the class directly in graf2d. I would not add the feynman folder but rather put the file in graf2d/graf/src and graf2d/graf/inc . You need then to modify graf2d/graf/inc/LinkDef.h

Okay, i'll move the files

**Comment by AdvaitDhingra** [comment]
Is everything okay with the documentation?

**Comment by couet** [comment]
Now, the way you do it you will have the TLatex and the TCurly etc ... primitives in the TCanvas list (try gPad->ls()). The object TFeynman does not exits on its own.  Why not having:

```
   auto f = new TFeynman(...);
   f->Add(Lepton(10, 10, 30, 30, 7, 6, "e", true));
   f->Add(Lepton(10, 50, 30, 30, 5, 55, "e", false));
   f->Add(CurvedPhoton(30, 30, 12.5*sqrt(2), 135, 225, 7, 30));
   f->Add(Photon(30, 30, 55, 30, 42.5, 37.7));
   f->Add(QuarkAntiQuark(70, 30, 15, 55, 45, "q"));
   f->Add(Gluon(70, 45, 70, 15, 77.5, 30)-);
   f->Add(WeakBoson(85, 30, 110, 30, 100, 37.5, "Z^{0}"));
   f->Add(Quark(110, 30, 130, 50, 135, 55, "q", true));
   f->Add(Quark(110, 30, 130, 10, 135, 6, "q", false));
   f->Add(CurvedGluon(110, 30, 12.5*sqrt(2), 315, 45, 135, 30));
   f->Draw()
```

The Lepton CurvedPhoton etc ... would be some TFeynmanObject (we should find a good name), a bit like TLegend and TLegendItem. TFeynman should inherits from TAttLine and TAttText to define the global line and text attributes of the diagram and TFeynmanObject should also inherits from TAttLine and TAttText  to change individual TFeynmanObject attributes. TFeynman will need a Draw(...) method and a Paint(...) method. The Paint(...) method will call all the Paint(...) methods of the TFeynmanObject building the TFeynman diagram.  You may also what to implement TFeynman::SavePrimitive. 

I also see that you need to define the position of the objects building the diagram. If you look at https://arxiv.org/pdf/1601.05437.pdf the diagram is defined without giving any absolute position. Would it be possible to do something similar ?

**Comment by AdvaitDhingra** [comment]
> Now, the way you do it you will have the TLatex and the TCurly etc ... primitives in the TCanvas list (try gPad->ls()). The object TFeynman does not exits on its own. Why not having:
> 
> ```
>    auto f = new TFeynman(...);
>    f->Add(Lepton(10, 10, 30, 30, 7, 6, "e", true));
>    f->Add(Lepton(10, 50, 30, 30, 5, 55, "e", false));
>    f->Add(CurvedPhoton(30, 30, 12.5*sqrt(2), 135, 225, 7, 30));
>    f->Add(Photon(30, 30, 55, 30, 42.5, 37.7));
>    f->Add(QuarkAntiQuark(70, 30, 15, 55, 45, "q"));
>    f->Add(Gluon(70, 45, 70, 15, 77.5, 30)-);
>    f->Add(WeakBoson(85, 30, 110, 30, 100, 37.5, "Z^{0}"));
>    f->Add(Quark(110, 30, 130, 50, 135, 55, "q", true));
>    f->Add(Quark(110, 30, 130, 10, 135, 6, "q", false));
>    f->Add(CurvedGluon(110, 30, 12.5*sqrt(2), 315, 45, 135, 30));
>    f->Draw()
> ```
> 
> The Lepton CurvedPhoton etc ... would be some TFeynmanObject (we should find a good name), a bit like TLegend and TLegendItem. TFeynman should inherits from TAttLine and TAttText to define the global line and text attributes of the diagram and TFeynmanObject should also inherits from TAttLine and TAttText to change individual TFeynmanObject attributes. TFeynman will need a Draw(...) method and a Paint(...) method. The Paint(...) method will call all the Paint(...) methods of the TFeynmanObject building the TFeynman diagram. You may also what to implement TFeynman::SavePrimitive.
> 
> I also see that you need to define the position of the objects building the diagram. If you look at https://arxiv.org/pdf/1601.05437.pdf the diagram is defined without giving any absolute position. Would it be possible to do something similar ?

Thanks for the tips. I will try to do this. 😄

**Comment by couet** [comment]
> Thanks for the tips. I will try to do this. 😄
Of course any comments or criticism on what I suggested are welcome :-)

**Comment by AdvaitDhingra** [comment]
> > Thanks for the tips. I will try to do this. 😄
> 
> 
> 
> Of course any comments or criticism on what I suggested are welcome :-)

Thanks. So if I understand correctly, instead of plotting the TLine, TArrow etc one adds them to the TFeynman object and then draws everything at once? I looked into the code for TLegend, and this is what I understand:

- The legend entries are added to a TList
- The Draw method takes an Option_p and appends it to the TPad. What is this option_p?

My apologies if these are simple questions

**Comment by couet** [comment]
The general structure would be something like [TLegend](https://root.cern/doc/master/classTLegend.html) . Of course an automatic positioning is more tricky in the TFeynman case.

**Comment by couet** [comment]
> Thanks. So if I understand correctly, instead of plotting the TLine, TArrow etc one adds them to the TFeynman object and > > then draws everything at once? I looked into the code for TLegend, and this is what I understand:

Yes I would define your objects Lepton etc .. add being TFeynmanItems

**Comment by AdvaitDhingra** [comment]
> Yes you are right .. that's an idea. II think it makes more senses ... no ?

Yeah, this approach seems more practical.

**Comment by AdvaitDhingra** [comment]
Hey, @couet, I'm working on the TFeynmanEntry class (which goes with the TFeynman class) and I was wondering how to make the Draw() function. This is how I have done it at the moment, but it tells me that `AppendPad()` is unidentified. What have I done wrong?  

``` c++
// TFeynman.cxx:
// imports:
#include <cstdio>
#include <iostream>

#include "TStyle.h"
#include "TLatex.h"
#include "TLine.h"
#include "TPolyLine.h"
#include "TMarker.h"
#include "../inc/TFeynman.h"
#include "TList.h"
#include "TVirtualPad.h"
#include "TMath.h"
#include "TROOT.h"
#include "TMultiGraph.h"
#include "TGraph.h"
#include "TH1.h"
#include "THStack.h"

// the function itself:
void TFeynman::Draw() {
	AppendPad();
}
// declaration of the function:
virtual void Draw();
```
What is wrong here? Sorry for bothering you.

**Comment by AdvaitDhingra** [comment]
I pushed the changes so you can see all of the files. It's not finished yet, as I'm having trouble with AppendPad() (see comment above)

**Comment by couet** [comment]
You do not have any Draw() method for the TFeynmanEntry. You have a Add() method which will add the new object in the list of object in TFeynman. So you need to have a Tlist in TFeynman holding the list of all the TFeynman entry buildingthe diagram. The Draw method of TFeynman will simply do AppendPad and will pout the whole diagram in the list of primates of the current pad (gPad) . TFeynmanEntry will have a Paint and the Paint method of TFeynman will call all the Paint of the TFeynmanEntry in the TList ... Really it is the same structure as TLegend.

**Comment by couet** [comment]
We can discuss it in zoom if you like

**Comment by AdvaitDhingra** [comment]
> We can discuss it in zoom if you like

Thank you for the offer. I think I'll try it myself first and if I get stuck I'll send you an email, ok?

**Comment by couet** [comment]
That's fine

**Comment by couet** [comment]
Lepton, CurvedPhoton, Photon, QuarkAntiQuark, Gluon, WeakBoson, Quark, CurvedGluon should not be methods of TFeynman but they should inherit from TFeynmanItem and from TAttLine and TAttText. The Add method of TFeynman then can be something like:

TFeynman::Add(TFeynman *item)

The you can have the macro as I suggested.

**Comment by AdvaitDhingra** [comment]
> Lepton, CurvedPhoton, Photon, QuarkAntiQuark, Gluon, WeakBoson, Quark, CurvedGluon should not be methods of TFeynman but they should inherit from TFeynmanItem and from TAttLine and TAttText. The Add method of TFeynman then can be something like:
> 
> TFeynman::Add(TFeynman *item)
> 
> The you can have the macro as I suggested.

Ah ok. So instead of having them as methods of the class, where would I define them?

**Comment by couet** [comment]
The code to Draw will be in the Paint() method of each FeynmanItem

**Comment by AdvaitDhingra** [comment]
> The code to Draw will be in the Paint() method of each FeynmanItem

Ahh ok, go it. Thanks

**Comment by couet** [comment]
> where would I define them?

You can define them in the TFeynman file no need to have other files.

**Comment by AdvaitDhingra** [comment]
> > where would I define them?
> 
> You can define them in the TFeynman file no need to have other files.

Ok thanks. My school is done now so I'll get started now.

**Comment by AdvaitDhingra** [comment]
I've sent you an email regarding a Zoom meeting.

**Comment by AdvaitDhingra** [comment]
Good Morning Mr. @couet . I just realized that the paint method of TFeynman is never being called. When I call the draw method, do I want to run the paint method first before ´AppendPad`?

**Comment by AdvaitDhingra** [comment]
In the TLegend::AddEntry method, a TLegendEntry is added to the TList (fPrimitives). When I try to do this, it tells me that the Add() method of TList takes in an Object and I provided a TFeynmanEntry . How did TLegend do this?

**Comment by couet** [comment]
> Good Morning Mr. @couet . I just realized that the paint method of TFeynman is never being called. When I call the draw > > > method, do I want to run the paint method first before ´AppendPad`?

The Paint() method is called automatically for you when the Pad will be display. You do not need to call it. See the figure in TPad ref guide .

**Comment by AdvaitDhingra** [comment]
> > Good Morning Mr. @couet . I just realized that the paint method of TFeynman is never being called. When I call the draw > > > method, do I want to run the paint method first before ´AppendPad`?
> 
> The Paint() method is called automatically for you when the Pad will be display. You do not need to call it. See the figure in TPad ref guide .

Ah ok. And when I want to add a TFeynmanEntry to the TList, how can I do that?

**Comment by couet** [comment]
> In the TLegend::AddEntry method, a TLegendEntry is added to the TList (fPrimitives). When I try to do this, it tells me that > > the Add() method of TList takes in an Object and I provided a TFeynmanEntry . How did TLegend do this?

Ah yes TFeynman and TFeynmanEntry should inherits from TObject like TLegendEntry. See the inheritance graph here:

https://root.cern/doc/master/classTLegendEntry.html

**Comment by AdvaitDhingra** [comment]
> > In the TLegend::AddEntry method, a TLegendEntry is added to the TList (fPrimitives). When I try to do this, it tells me that > > the Add() method of TList takes in an Object and I provided a TFeynmanEntry . How did TLegend do this?
> 
> Ah yes TFeynman and TFeynmanEntry should inherits from TObject like TLegendEntry. See the inheritance graph here:
> 
> https://root.cern/doc/master/classTLegendEntry.html

Ahhhh okay. This makes sense. Thank you

**Comment by AdvaitDhingra** [comment]
It's starting to look much better now :)

**Comment by AdvaitDhingra** [comment]
For some reason it automatically requested a review from someone else. Strange

**Comment by couet** [comment]
> For some reason it automatically requested a review from someone else. Strange

No worry it is normal. That's because you changed the makefiles . Each part of root is assigned to somebody. The Make part is assigned to Oksana. That's why she was automatically asked. But we do not need to bother her with that. That's why I removed the request review.

**Comment by AdvaitDhingra** [comment]
> > For some reason it automatically requested a review from someone else. Strange
> 
> 
> 
> No worry it is normal. That's because you changed the makefiles . Each part of root is assigned to somebody. The Make part is assigned to Oksana. That's why she was automatically asked. But we do not need to bother her with that. That's why I removed the request review.

Ahh ok

**Comment by AdvaitDhingra** [comment]
When I try to iterate over all of the TFeynmanEntry 's in fPrimitives (a TList), and call the Paint() Method of that TFeynmanEntry, I get a segmentation viloation: 

```
 *** Break *** segmentation violation



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00007f07a687fdba in __GI___wait4 (pid=172350, stat_loc=stat_loc
entry=0x7ffdbec66c68, options=options
entry=0, usage=usage
entry=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:27
#1  0x00007f07a687fd7b in __GI___waitpid (pid=<optimized out>, stat_loc=stat_loc
entry=0x7ffdbec66c68, options=options
entry=0) at waitpid.c:38
#2  0x00007f07a67ef0e7 in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:172
#3  0x00007f07a6eb1afe in TUnixSystem::StackTrace() () from /home/advait/rootbuild/lib/libCore.so
#4  0x00007f07a6eae985 in TUnixSystem::DispatchSignals(ESignals) () from /home/advait/rootbuild/lib/libCore.so
#5  <signal handler called>
#6  0x0000000000000061 in ?? ()
#7  0x00007f07a6b8d6fb in ?? ()
#8  0x0000000000000000 in ?? ()
===========================================================


The lines below might hint at the cause of the crash.
You may get help by asking at the ROOT forum https://root.cern.ch/forum
Only if you are really convinced it is a bug in ROOT then please submit a
report at https://root.cern.ch/bugs Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#6  0x0000000000000061 in ?? ()
#7  0x00007f07a6b8d6fb in ?? ()
#8  0x0000000000000000 in ?? ()
===========================================================

```
Here is the code for looping over the TFeynmanEntry 's:
```
void TFeynman::Paint() {
	cout << "Paint Method called. Grab your paintbrush" << endl;
	TIter next(fPrimitives);
	TFeynmanEntry *entry;
	const char* particle = entry->GetParticleName();
	while (( entry = (TFeynmanEntry*)next() )) {
		cout << "Iter" << endl;
	}
```
TLegend does it basically the same way (at least the loop, TLegend doesn't call the Paint() Method of TLegendEntry). Is my PC incapable of handling the loop or is it a code problem?

**Comment by AdvaitDhingra** [comment]
Oh my god what just happend

**Comment by AdvaitDhingra** [comment]
Sorry for that. I must have messed up the commits somehow. Luckily I use Atom, so my files were not lost. Sorry for that

**Comment by AdvaitDhingra** [comment]
Thank you @couet . I'm not sure what happened but I managed to fix the commits.

**Comment by couet** [comment]
Ok I do not see changes compare to the last comments I did. I suggested you now modify the test example according to the new structure and try if it works.

**Comment by AdvaitDhingra** [comment]
> Ok I do not see changes compare to the last comments I did. I suggested you now modify the test example according to the new structure and try if it works.

Okay, I will do it. Do I need to build root first? If yes, I am unsure if I structured LinkDef.h correctly. Is is done correctly?

**Comment by couet** [comment]
> Do I need to build root first? If yes, I am unsure if I structured LinkDef.h correctly. Is is done correctly?

Yes of course. You need to compile you class . You need to type "make" in the place you have you root build.

**Comment by AdvaitDhingra** [comment]
I built it and the build works fine. The only problem is that the Paint() method is not being called automatically. Did I forget to import something, because i imported TVirtualPad.h .

**Comment by AdvaitDhingra** [comment]
When i call the Paint() method of TFeynman manually in the feynman.C file it still doesn't display anyting.

**Comment by couet** [comment]
What is your example macro ?

**Comment by AdvaitDhingra** [comment]
> What is your example macro ?

The example I made is in srs/feynman.C . I ran it on the version of ROOT that I buit frm source and it works. except that it doesn't call the Paint() method automatically when Draw() is called.

And when I call Paint() manually it still does not display anything.

**Comment by couet** [comment]
I see you macro. it should be:

```
void feynman()
 {
     TCanvas *c1 = new TCanvas();
     TFeynman *f = new TFeynman();
     f->AddItem("fermion", 10, 10, 30, 30);
     f->Draw();
 }
```

then you start you new built root. And do:
```
root[0] .x feynman.C
```
is it what you are doing ?

**Comment by AdvaitDhingra** [comment]
> I see you macro. it should be:
> 
> ```
> void feynman()
>  {
>      TCanvas *c1 = new TCanvas();
>      TFeynman *f = new TFeynman();
>      f->AddItem("fermion", 10, 10, 30, 30);
>      f->Draw();
>  }
> ```
> 
> then you start you new built root. And do:
> 
> ```
> root[0] .x feynman.C
> ```
> 
> is it what you are doing ?

Actually I created a buildroot directory outside of the root directory on my laptop and I run 
```
cmake -DCMAKE_INSTALL_PREFIX=../rootinstall ../root

cmake --build . -- -j4
```
And then once it is done I navigate to the buildroot/bin directory and run 
```
source thisroot.sh
./root test.C
```
The test.C has the same contents as feynman.C

**Comment by couet** [comment]
I am building with your new classes ... I will try

**Comment by couet** [comment]
Ok I have it
```
root [0] TCanvas *c1 = new TCanvas();
root [1] TFeynman *f = new TFeynman();
root [2] f->Draw();
Draw Method called. Grab your pencils.
Paint Method called. Grab your paintbrush
root [3] gPad->ls()
Canvas Name=c1 Title=c1 Option=
 TCanvas fXlowNDC=0 fYlowNDC=0 fWNDC=1 fHNDC=1 Name= c1 Title= c1 Option=
  OBJ: TList	TList	Doubly linked list : 0
   OBJ: TFeynman	TFeynman	Feynman diagram : 0 at: 0x7f8b2d8a1350
root [4] 

```

**Comment by couet** [comment]
I send you the modified files by email

**Comment by AdvaitDhingra** [comment]
Ah, perfect. Thank you so much for your help. I will get started on the other particles aswell.

**Comment by AdvaitDhingra** [comment]
I added your name to everything

**Comment by AdvaitDhingra** [comment]
Is it just me, or are the particles not visible on the TPad? I tried adding a Fermion but the TPad is blank.

**Comment by AdvaitDhingra** [comment]
> Is it just me, or are the particles not visible on the TPad? I tried adding a Fermion but the TPad is blank.

Nevermind, you just need to add 
```
c1->Range(0, 0, 140, 60);
```
Then it works

**Comment by AdvaitDhingra** [comment]
Now it automatically fixes this issue.

**Comment by AdvaitDhingra** [comment]
I just saw that TCurlyLine has no Paint method. Is there an equivalent?

**Comment by AdvaitDhingra** [comment]
> ## DeepCode's analysis on [#32656b](https://www.deepcode.ai/app/gh/root-project/root/67ef1b3ab32ff72f82707da58e71bc0c702f51b4/root-project/root/32656bcc40899d2945eecfccf6710dd98407d7a7/pr/_/%2F/code/?utm_source=gh_review&c=0&w=1&i=0&) found:
> * warning **1** warning point_down
> 
> ## Top issues
> Description	Example fixes
> Leaking memory. TArrow is allocated on the heap and never freed. In function Paint. Occurrences:
> * [TFeynmanEntry.cxx:42](https://github.com/root-project/root/blob/32656bcc40899d2945eecfccf6710dd98407d7a7/graf2d/graf/src/TFeynmanEntry.cxx#L42)
> * [TFeynmanEntry.cxx:45](https://github.com/root-project/root/blob/32656bcc40899d2945eecfccf6710dd98407d7a7/graf2d/graf/src/TFeynmanEntry.cxx#L45)
> * [TFeynmanEntry.cxx:49](https://github.com/root-project/root/blob/32656bcc40899d2945eecfccf6710dd98407d7a7/graf2d/graf/src/TFeynmanEntry.cxx#L49)
> 
> [wrench Example fixes](https://www.deepcode.ai/app/gh/root-project/root/67ef1b3ab32ff72f82707da58e71bc0c702f51b4/root-project/root/32656bcc40899d2945eecfccf6710dd98407d7a7/pr/_/%2Fgraf2d%2Fgraf%2Fsrc%2FTFeynmanEntry.cxx/cpp%2Fdc%2FCppMemoryLeak/code/?utm_source=gh_review&)
> ####  View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/root-project/root/67ef1b3ab32ff72f82707da58e71bc0c702f51b4/root-project/root/32656bcc40899d2945eecfccf6710dd98407d7a7/pr/_/%2F/code/?utm_source=gh_review&c=0&w=1&i=0&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=root-project)

How can this be fixed? The method mentioned does not seem the best

**Comment by couet** [comment]
> I just saw that TCurlyLine has no Paint method. Is there an equivalent?

Look at the inheritance ... it gets it from TPolyline.

**Comment by AdvaitDhingra** [comment]
Is it okay if I add my name to root/README/CREDITS?

**Comment by couet** [comment]
> Is it okay if I add my name to root/README/CREDITS?

May be when/if the class will be put in official root ? that's what we do usually. We add in CREDITS when a fully working product has been provided.

**Comment by couet** [comment]
Can you send me a tar file by email with what you did already so I can easily try it ?

**Comment by AdvaitDhingra** [comment]
> > Is it okay if I add my name to root/README/CREDITS?
> 
> 
> 
> May be when/if the class will be put in official root ? that's what we do usually. We add in CREDITS when a fully working product has been provided.

Ahh okay. What do you think of the particle- antiparticle functionality. It places the labels automatically.

**Comment by couet** [comment]
send me the code so I can try ...

**Comment by AdvaitDhingra** [comment]
> send me the code so I can try ...

It's part of the last two commits

**Comment by AdvaitDhingra** [comment]
> send me the code so I can try ...

Sorry, I saw this a little late. I sent the email now.

**Comment by AdvaitDhingra** [comment]
Good Morning @couet . I have a question regarding the Paint() method of TFeynmanEntry. It is telling me that the parameter Option_t* option is unused. While this is not causing any errors, I was wondering if the parameter needs to be used in the method somewhere?

**Comment by AdvaitDhingra** [comment]
@couet good morning. Do you know how I can fix the deepcode issue?

**Comment by couet** [comment]
> @couet good morning. Do you know how I can fix the deepcode issue?

No, I cannot see what it is about when I click on "Details"

**Comment by AdvaitDhingra** [comment]
> > @couet good morning. Do you know how I can fix the deepcode issue?
> 
> No, I cannot see what it is about when I click on "Details"

It is saying that the TArrow from TFeynmanEntry is not being freed from the heap

**Comment by couet** [comment]
> > > @couet good morning. Do you know how I can fix the deepcode issue?
> > 
> > 
> > No, I cannot see what it is about when I click on "Details"
> 
> It is saying that the TArrow from TFeynmanEntry is not being freed from the heap

So free it ...

**Comment by AdvaitDhingra** [comment]
It turns out this fix was needed in all particles. Thank you :)

**Comment by AdvaitDhingra** [comment]
I think almost everything is there now. I've tested it in both Ubuntu 20.04 LTS and Windows 10. Can we let @phsft-bot build it on other systems?

**Comment by couet** [comment]
> I think almost everything is there now. I've tested it in both Ubuntu 20.04 LTS and Windows 10. Can we let @phsft-bot build it on other systems?

No, it is not, see my comments about the coordinates ... how to you define them ?

**Comment by AdvaitDhingra** [comment]
> > I think almost everything is there now. I've tested it in both Ubuntu 20.04 LTS and Windows 10. Can we let @phsft-bot build it on other systems?
> 
> No, it is not, see my comments about the coordinates ... how do you define them ?

I added Doxygen comments on the parameters of the methods. The coordinates are like this:

For AddItem you provide x1, y1 which is the starting coordinate and x2, y2 which is the stopping coordinate and labelX and labelY whihc is where the label is. I'm trying to figure out a way where you don't need to set the coordinates, but that may take a while.

For AddPair, you simply provide the x and y coordinate of the center of the arc. The labels are positioned automatically.

For AddCurved, You also only provide the x and y coordinate of the center of the arc, along with phimin and phimax angles. I'm also trying to find a way to reduce the amount of parameters here (like in TikZ).

The problem is that the main type of feynman Diagram, is one where you have the fermions at the left and right side, and a boson + virtual particles (in pairs) in the middle. If this were the only type of feynman diagram, I could do this easily, however sometimes there are other "irregular" diagrams, which would be harder to implement.

Is there a way, where the particles can be positioned automatically, and then checked to make sure there is no overlapping?

**Comment by couet** [comment]
I mean the coordinates you are using are not defined . What is the Max ? what is the min ? in which space are you drawing ? try the the following example. Where is the diagram ?

```
#include "TFeynman.h"
#include "TFeynmanEntry.h"

void feynman()
{
  TCanvas *c1 = new TCanvas("c1","c1",800,800);

  auto g = new TGraph();
  g->AddPoint(1,2);
  g->AddPoint(2,3);
  g->AddPoint(3,1);
  g->Draw("AL*");

  auto f = new TFeynman();
  // proton decay (beta minus)
  //f->AddPair("q", 50, 30, 6);
  f->AddItem("fermion", 10, 10, 30, 30, 5, 6, "d");
  f->AddItem("fermion", 30, 30, 10, 50, 5, 50, "d");
  f->AddItem("fermion", 15, 10, 35, 30, 10, 6, "u");
  f->AddItem("fermion", 35, 30, 15, 50, 12, 50, "u");
  f->AddItem("fermion", 20, 10, 40, 30, 15, 6, "u");
  f->AddItem("fermion", 40, 30, 20, 50, 17, 50, "d");
  f->AddItem("boson", 40, 30, 70, 30, 55, 35, "W^{+}");
  f->AddItem("anti-fermion", 70, 30, 90, 50, 95, 55, "e^{+}");
  f->AddItem("fermion", 70, 30, 90, 10, 85, 5, "#bar{#nu}");
  f->AddCurved("#gamma", 80, 30, 12, 275, 90, true);
  f->AddPair("u", 55, 30, 10);
  f->Draw();
}

```

**Comment by couet** [comment]
I have a solution in mind ... may be you thought of it already ?  What about making TFeynman a TPad ?

**Comment by AdvaitDhingra** [comment]
> I have a solution in mind ... may be you thought of it already ? What about making TFeynman a TPad ?

Ah, this is a good idea. So the TFeynman would contain TFeynmanEntries (as it does right now), and these would then be positioned and painted automatically? This sounds doable. Is this what you have in mind?

**Comment by AdvaitDhingra** [comment]
> I mean the coordinates you are using are not defined . What is the Max ? what is the min ? in which space are you drawing ? try the the following example. Where is the diagram ?
> 
> ```
> #include "TFeynman.h"
> #include "TFeynmanEntry.h"
> 
> void feynman()
> {
>   TCanvas *c1 = new TCanvas("c1","c1",800,800);
> 
>   auto g = new TGraph();
>   g->AddPoint(1,2);
>   g->AddPoint(2,3);
>   g->AddPoint(3,1);
>   g->Draw("AL*");
> 
>   auto f = new TFeynman();
>   // proton decay (beta minus)
>   //f->AddPair("q", 50, 30, 6);
>   f->AddItem("fermion", 10, 10, 30, 30, 5, 6, "d");
>   f->AddItem("fermion", 30, 30, 10, 50, 5, 50, "d");
>   f->AddItem("fermion", 15, 10, 35, 30, 10, 6, "u");
>   f->AddItem("fermion", 35, 30, 15, 50, 12, 50, "u");
>   f->AddItem("fermion", 20, 10, 40, 30, 15, 6, "u");
>   f->AddItem("fermion", 40, 30, 20, 50, 17, 50, "d");
>   f->AddItem("boson", 40, 30, 70, 30, 55, 35, "W^{+}");
>   f->AddItem("anti-fermion", 70, 30, 90, 50, 95, 55, "e^{+}");
>   f->AddItem("fermion", 70, 30, 90, 10, 85, 5, "#bar{#nu}");
>   f->AddCurved("#gamma", 80, 30, 12, 275, 90, true);
>   f->AddPair("u", 55, 30, 10);
>   f->Draw();
> }
> ```

Ahh ok. I'll take a look

**Comment by couet** [comment]
At the end it might be better to submit your classes as a macro example in the forum so people can try and give feed back. I do not think it is mature enough to be a native ROOT class. I am sure people will have many criticism and new ideas to propose.

**Pull request closed without merging**